pkg/services/dashboards/service/zanzana.go

```diff
@@ -0,0 +1,342 @@
+					dr.log.Error("error checking access", "error", err)
+				} else if allowed {
+					allowedResults <- d
+				}
+			}
+		}()
+	}
+
+	for _, r := range searchRes {
+		resToCheck <- r
+	}
+	close(resToCheck)
+
+	wg.Wait()
+	close(allowedResults)
+
+	for r := range allowedResults {
+		if int64(len(res)) >= remains {
+			break
+		}
+		res = append(res, r)
+	}
+
+	return res, nil
+}
+
+// findDashboardsZanzanaList implements "List, then search" strategy. It first retrieve a list of resources
+// with given type available to the user and then passes that list as a filter to the search query.
+func (dr *DashboardServiceImpl) findDashboardsZanzanaList(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
+	// Always use "search, then check" if dashboard or folder UIDs provided. Otherwise we should make intersection
+	// of user's resources and provided UIDs which might not be correct if ListObjects() request is limited by OpenFGA.
+	if len(query.DashboardUIDs) > 0 || len(query.DashboardIds) > 0 || len(query.FolderUIDs) > 0 {
+		return dr.findDashboardsZanzanaCheck(ctx, query)
+	}
+
+	ctx, span := tracer.Start(ctx, "dashboards.service.findDashboardsZanzanaList")
+	defer span.End()
+
+	resourceUIDs, err := dr.listUserResources(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	if len(resourceUIDs) == 0 {
+		return []dashboards.DashboardSearchProjection{}, nil
+	}
+
+	query.DashboardUIDs = resourceUIDs
+	query.SkipAccessControlFilter = true
+	return dr.dashboardStore.FindDashboards(ctx, query)
+}
+
+func (dr *DashboardServiceImpl) listUserResources(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery) ([]string, error) {
+	tasks := make([]func() ([]string, error), 0)
+	var resourceTypes []string
+
+	// For some search types we need dashboards or folders only
+	switch query.Type {
+	case searchstore.TypeDashboard:
+		resourceTypes = []string{zanzana.TypeDashboard}
+	case searchstore.TypeFolder, searchstore.TypeAlertFolder:
+		resourceTypes = []string{zanzana.TypeFolder}
+	default:
+		resourceTypes = []string{zanzana.TypeDashboard, zanzana.TypeFolder}
+	}
+
+	for _, resourceType := range resourceTypes {
+		tasks = append(tasks, func() ([]string, error) {
+			return dr.listAllowedResources(ctx, query, resourceType)
+		})
+	}
+
+	uids, err := runBatch(tasks)
+	if err != nil {
+		return nil, err
+	}
+
+	return uids, nil
+}
+
+func (dr *DashboardServiceImpl) listAllowedResources(ctx context.Context, query *dashboards.FindPersistedDashboardsQuery, resourceType string) ([]string, error) {
+	res, err := dr.ac.ListObjects(ctx, accesscontrol.ListObjectsRequest{
+		User:     query.SignedInUser.GetUID(),
+		Type:     resourceType,
+		Relation: "read",
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	orgId := query.OrgId
+	if orgId == 0 && query.SignedInUser.GetOrgID() != 0 {
+		orgId = query.SignedInUser.GetOrgID()
+	} else {
+		return nil, dashboards.ErrUserIsNotSignedInToOrg
+	}
```

Comment:
Something looks off here, what are we trying to control? That the user org is the same as the query org?

Reply:
We might have org id set as a query param, so it might be different from user's org where it logged in. It's copy from search query:

https://github.com/grafana/grafana/blob/babddc9bf254ac8d9477c342f776bdbee0dcb73f/pkg/services/dashboards/database/database.go#L901-L905